### PR TITLE
[CSRanking] Favour concrete members over protocol requirements

### PR DESCRIPTION
**Pull request by hamishknight**
This PR comes in two pieces:

#### 1. A narrow Swift 4.1 compatibility hack that favours properties in concrete types over protocol requirements

This was unintentionally regressed by #15412, and is tracked by [SR-7425]. This is added as a ranking tie-breaker in order to ensure we don’t introduce any new ambiguities.

#### 2. A new Swift 5 mode ranking rule that favours arbitrary concrete type members over protocol requirements

We already have a similar rule that prefers concrete members over protocol extension members – IMO this new rule is a natural extension of it.

This allows the following code to become unambiguous in Swift 5 mode:

```swift
protocol X {
  func baz() -> Int
}

class Y {
  func baz() -> Int { return 0 }
}

func foo(_ x: X & Y) {
  // currently ambiguous, will be made unambigous in Swift 5 mode (favouring Y's `baz()`)
  x.baz()
}
```

Note that this new rule isn’t applied to overloads found through dynamic lookup in order to ensure we keep cases such as [SR-7299] ambiguous.

Note also that this new rule is potentially source breaking (hence why it requires Swift 5 mode) – for example, the following contrived case is currently unambiguous, but will be made ambiguous by the change:

```swift
protocol P {
  static func foo() -> Int
}

class C {
  func foo() -> Int { return 0 }
}

extension P where Self : C {
  static func bar() {
    let m = foo // currently unambiguous, will be made ambiguous in Swift 5 mode
  }
}
```

Arguably this change is unifying the behaviour with that of protocol extensions, where a similar example is already ambiguous:

```swift
protocol P {}
extension P {
  static func foo() -> Int { return 0 }
}

class C {
  func foo() -> Int { return 0 }
}

extension P where Self : C {
  static func bar() {
    let m = foo // error: Ambiguous use of 'foo()'
  }
}
```

And currently, adding `static func foo()` as a protocol requirement to the above suddenly resolves the ambiguity – with this Swift 5 mode change, the ambiguity is kept in both cases, which seems more consistent.

I’m happy to split this PR up if the Swift 5 mode change requires more discussion.

Resolves [SR-7425].

  [SR-7425]: https://bugs.swift.org/browse/SR-7425
  [SR-7299]: https://bugs.swift.org/browse/SR-7299

### CODE REVIEW

**Comment by hamishknight** [comment]
Just fixed the Linux failure. The source compatibility suite succeeded, though looking [through the projects](https://github.com/apple/swift-source-compat-suite/blob/master/projects.json) I realise that we wouldn't have exercised `-swift-version 5` at all. Is there a way to do so?

**Comment by rudkx** [comment]
@hamishknight No. What I would suggest is removing the version check for the sake of running the compat suite to check that there aren't any unexpected breaks (and see what actually will break for people moving to Swift 5).

**Comment by hamishknight** [comment]
@rudkx Thanks, I've gone ahead and done that – could you kick off the compat suite again?

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
Can you split the changes up into separate PRs we get some feedback from the source compatibility suite?

I'll take a closer look at each once that's done. I am a little concerned that at first glance some of the changes remind me of https://github.com/apple/swift/pull/17995

**Comment by hamishknight** [comment]
Sure thing – I've split it up into #18952 & #18951. Shall I go ahead and close this PR?

File: lib/Sema/CSRanking.cpp [hunk 1]
```diff
@@ -446,6 +451,32 @@ static bool isDeclAsSpecializedAs(TypeChecker &tc, DeclContext *dc,
...
+      //      var i: String { get }
+      //    }
+      //    class C {
+      //      @objc var i: Int { return 0 }
+      //    }
+      //    func foo(_ x: AnyObject) {
+      //      x.i // ensure ambiguous.
+      //    }
+      //
+      if (true && !isDynamicOverloadComparison) {
```
Comment (xedin) [inline lib/Sema/CSRanking.cpp]: It should probably be simplified to `if (!isDynamicOverloadComparison)`

File: lib/Sema/CSRanking.cpp [hunk 2]
```diff
@@ -962,6 +1004,30 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
...
+    // type, which created unintentional ambiguities in 4.2. This hack ensures
+    // we at least keep these cases unambiguous in Swift 5 under Swift 4
+    // compatibility mode. Don't however apply this hack for decls found through
+    // dynamic lookup, as we want the user to have to disambiguate those.
+    //
+    // This is intentionally narrow in order to best preserve source
+    // compatibility under Swift 4 mode by ensuring we don't introduce any new
+    // ambiguities. This will become a more general "is more specialised" rule
+    // in Swift 5 mode.
+    if (false && !isDynamicOverloadComparison &&
```
Comment (xedin) [inline lib/Sema/CSRanking.cpp]: Does this mean that this code shouldn't exist?

File: lib/Sema/CSRanking.cpp [hunk 3]
```diff
@@ -1118,6 +1184,14 @@ SolutionCompareResult ConstraintSystem::compareSolutions(
     score1 += isStdlibOptionalMPlusOperator2;
   }
 
+  // All other things being equal, apply the Swift 4 compatibility hack for
+  // preferring var members in concrete types over a protocol requirement
+  // (see the comment above for the rationale of this hack).
+  if (false && score1 == score2) {
```
Comment (xedin) [inline lib/Sema/CSRanking.cpp]: Same here?

File: lib/Sema/CSRanking.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/CSRanking.cpp]: @xedin I removed the version checks temporarily to allow the compat suite to exercise the Swift 5 path in 685ec32919d6c2efa4cbe9219d9c50857bd61482. `true` should be `tc.Context.isSwiftVersionAtLeast(5)`, and below, `false` should be `!tc.Context.isSwiftVersionAtLeast(5)`.

**Comment by xedin** [comment]
@rudkx As this branch stands today it preserves behavior when it comes to protocol vs. protocol extension declarations, and would always pick declaration on protocol...

File: lib/Sema/CSRanking.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSRanking.cpp]: Ah, gotcha!

**Comment by rudkx** [comment]
@hamishknight Yes, we can close this out in favor of the two new PRs. Thanks!

**Pull request closed without merging**